<launch>

    <arg name="use_kimera_frame" default="true"/>
    <arg name="use_oriented_bounding_boxes" default="true"/>
    <arg name="exit_after_bag" default="false"/>

    <arg name="enable_dsg_lcd" default="false"/>

    <arg name="dsg_output_dir" default="$(find kimera_dsg_builder)/output/sidpac"/>
    <arg name="dsg_output_prefix" default="sidpac_f1_3"/>
    <arg name="dsg_output_use_date" default="false"/>

    <arg name="dsg_should_log" default="true" />
    <arg name="pgmo_should_log" default="true"/>

    <arg name="semantic_map_dir" default="$(find semantic_recolor)/config/colors"/>
    <arg name="semantic_map_file" default="ade150_config.csv"/>
    <arg name="semantic_map_path" default="$(arg semantic_map_dir)/$(arg semantic_map_file)"/>

    <include file="$(find kimera_topology)/launch/kimera_topology_sidpac.launch">
        <arg name="use_kimera_frame" value="$(arg use_kimera_frame)"/>
        <arg name="show_stats" value="false"/>
    </include>

    <arg name="debug" default="false"/>
    <arg name="launch_prefix" value="gdb -ex run --args" if="$(arg debug)"/>
    <arg name="launch_prefix" value="" unless="$(arg debug)"/>

    <arg name="min_glog_level" default="3"/>
    <arg name="verbosity" default="0"/>
    <arg name="glog_to_file" default="false"/>
    <arg name="glog_dir" default="$(find kimera_dsg_builder)/glogs"/>
    <arg name="glog_file_args" value="--logtostderr=0 --log_dir=$(arg glog_dir)" if="$(arg glog_to_file)"/>
    <arg name="glog_file_args" value="" unless="$(arg glog_to_file)"/>

    <include file="$(find kimera_dsg_visualizer)/launch/includes/kimera_dsg_visualizer_params.xml">
        <arg name="viz_config_dir" value="$(find kimera_dsg_builder)/config/incremental_visualizer"/>
    </include>

    <rosparam command="load"
              file="$(find kimera_dsg_builder)/config/incremental_visualizer/agent_layer.yaml"
              ns="/kimera_dsg_visualizer/config/dynamic_layer/2"/>

    <arg name="dsg_path" value="$(arg dsg_output_dir)/$(arg dsg_output_prefix)"/>
    <arg name="pgmo_log_path" default="$(arg dsg_path)/pgmo"/>
    <node pkg="kimera_dsg_builder" type="kimera_dsg_builder_incremental_node"
          name="incremental_dsg_builder_node"
          launch-prefix="$(arg launch_prefix)"
          args="--minloglevel=$(arg min_glog_level) -v=$(arg verbosity) $(arg glog_file_args)"
          required="true"
          output="screen">
        <env name="OMP_NUM_THREADS" value="12"/>

        <param name="pgmo/log_path" value="$(arg pgmo_log_path)"/>
        <param name="pgmo/log_output" value="$(arg pgmo_should_log)"/>
        <rosparam file="$(find kimera_dsg_builder)/config/sidpac/dsg_frontend_config.yaml"/>
        <rosparam file="$(find kimera_dsg_builder)/config/sidpac/dsg_backend_config.yaml"/>
        <rosparam file="$(find kimera_dsg_builder)/config/sidpac/dsg_lcd_config.yaml" ns="lcd"/>


        <param name="exit_after_bag" value="$(arg exit_after_bag)"/>
        <param name="use_oriented_bounding_boxes" value="$(arg use_oriented_bounding_boxes)"/>
        <param name="semantic_label_2_color_csv_filepath" value="$(arg semantic_map_path)"/>

        <param name="dsg/output_path" command="date +'$(arg dsg_path)_%Y_%m_%d_%I_%M_%S.json'"
               if="$(arg dsg_output_use_date)"/>
        <param name="dsg/output_path" value="$(arg dsg_path).json"
               unless="$(arg dsg_output_use_date)"/>

        <param name="enable_lcd" value="$(arg enable_dsg_lcd)"/>

        <rosparam param="dynamic_semantic_labels">[20]</rosparam>
        <!-- ignoring 1, -->
        <rosparam param="structure_labels">[0, 1, 2, 3, 4, 5, 6, 7]</rosparam>
        <rosparam param="object_labels">[8, 9, 10, 11, 12, 13, 14, 15, 16]</rosparam>
        <remap from="~voxblox_mesh" to="/kimera_topology_node/mesh"/>
        <remap from="~active_places" to="/kimera_topology_node/active_layer"/>
        <remap from="~pose_graph_incremental" to="/kimera_vio_ros/pose_graph_incremental"/>
        <remap from="~/bow_vectors" to="/kimera_vio_ros/bow_query"/>
        <remap from="frame_registration" to="/kimera_vio_ros/kimera_vio_ros_node/register_lcd_frames"/>

        <rosparam file="$(find kimera_dsg_builder)/config/sidpac/place_mesh_graph_viz.yaml"
                  ns="pm_graph_viz"/>

        <param name="dsg_log_output" value="$(arg dsg_should_log)" />
        <param name="dsg_output_path" value="$(arg dsg_path)"/>
    </node>

</launch>
